test/e2e/data/infrastructure-docker/v1alpha4/bases/md.yaml

```diff
@@ -31,6 +31,11 @@ kind: MachineDeployment
 metadata:
   name: "${CLUSTER_NAME}-md-0"
 spec:
+  strategy:
+    type: "RollingUpdate"
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 0
```

Comment:
Shouldn't these be defaulted if they're not there? Does the test fail if these are not filled in?

Reply:
yeh they should, I just added it to test. Let me dig if/how this tests are running webhooks.

Reply:
cc @sbueringer I think you might be more familiar with this e2e tests, does this ring any bell, is there any known issue with webhooks?

Reply:
@enxebre I wasn't aware of it but looks like the docker infra provider "webhook" yaml is missing a lot of webhook configs: https://github.com/kubernetes-sigs/cluster-api/blob/master/test/infrastructure/docker/config/webhook/manifests.yaml

Reply:
@enxebre The error you had before imho proofs that at least the validating webhook was active:
```bash
1 --flavor (default)
INFO: Applying the cluster template yaml to the cluster
error: error validating "STDIN": error validating data: ValidationError(MachineDeployment.spec): missing required field "strategy" in io.x-k8s.cluster.v1alpha4.MachineDeployment.spec; if you choose to ignore these errors, turn validation off with --validate=false
```
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4527/pull-cluster-api-e2e-main/1387346828009148416/build-log.txt

Reply:
It looks like it should be deployed: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4527/pull-cluster-api-e2e-main/1387346828009148416/artifacts/repository/cluster-api/v0.4.0/components.yaml

Reply:
Thanks! Do you know if we have any logs for the webhook pod?

Reply:
I managed to reproduce this locally. Defaulting webhooko is failing consistently not only for machineDeployments.
This was probably failing silently Sincs this is possibly the first case we require a field and expect it to be defaulted.

```
E0429 11:37:22.071300       1 controller.go:302] controller-runtime/manager/controller/clusterresourceset "msg"="Reconciler error" "error"="Internal error occurred: failed calling webhook \"default.clusterresourceset.addons.cluster.x-k8s.io\": Post \"https://capi-webhook-service.capi-system.svc:443/mutate-addons-cluster-x-k8s-io-v1alpha4-clusterresourceset?timeout=10s\": dial tcp 10.96.132.19:443: connect: connection refused" "name"="kcp-adoption-73htjj-crs-0" "namespace"="kcp-adoption-46pnx5" "reconciler group"="addons.cluster.x-k8s.io" "reconciler kind"="ClusterResourceSet"
```

Reply:
I think with `failurePolicy`: `Fail` consistent failures should actually lead to completely blocked requests. And In my experience they do (and there's a bunch of conformance tests to ensure that).

As soon as the WebhookConfiguration is deployed there shouldn't be any race conditions.

There's definitely something wrong, though.

Reply:
Let's prioritize fixing the tests to always use the webhooks. Webhook code is treated as part of the reconciliation loop, we should make sure webhooks are always running.

Reply:
@enxebre I'll try you're branch locally. Maybe I find out what's going wrong.

Reply:
@enxebre I think I found the problem... Wouldn't have expected this:

It looks like the CRD validation is (also?) done client-side. After taking another look the error is pretty straightforward :):
```bash
error: error validating "STDIN": error validating data: ValidationError(MachineDeployment.spec): missing required field "strategy" in io.x-k8s.cluster.v1alpha4.MachineDeployment.spec; if you choose to ignore these errors, turn validation off with --validate=false
```

```bash
$ kubectl apply --help
...
--validate=true: If true, use a schema to validate the input before sending it
```

So while it failed with regular `kubectl apply` locally it worked without a problem when using `--validate=false`. It also looks like the defaulting worked flawlessly:
```bash
    strategy:
      rollingUpdate:
        maxSurge: 1
        maxUnavailable: 0
      type: RollingUpdate
```

While I'm not sure what we want to do now. We could hand over `validate=false` via `Args` here:
https://github.com/kubernetes-sigs/cluster-api/blob/7478817225e0a75acb6e14fc7b438231578073d2/test/e2e/quick_start.go#L75
in the `clusterctl.ApplyClusterTemplateAndWaitInput` struct.

Reply:
>It looks like the CRD validation is (also?) done client-side. After taking another look the error is pretty straightforward :)

Right that makes sense @sbueringerl, thanks.

>While I'm not sure what we want to do now. We could hand over validate=false via Args here:

I think that'd be unfortunate since every consumer/user using e.g kubectl would see the same behaviour which kinda defeat the purpose of our defaulting.

I would have expected open API kubectl client side validation to account also for structural schema defaults, but it seems that's no the case either. So I'm not sure now it's possible to make the field required while keeping the same UX for defaulting.

Reply:
I think you're right. As we cannot make it required this way without enforcing that it's already set on the client-side, we can only do it in the validating webhook. The validating webhook (Validator) would then basically validate if the mutating webhook (Defaulter) works correctly.

I assume this whole situation changes if we drop the pointer on these fields.